src/utils.rs

```diff
@@ -144,6 +144,23 @@ pub fn make_relative_path(base: &str, target: &str) -> String {
     }
 }
 
+pub fn greatest_lower_bound<'a, T, K: Ord, F: Fn(&'a T) -> K>(
+    slice: &'a [T],
+    key: &K,
+    map: F,
+) -> Option<&'a T> {
+    match slice.binary_search_by_key(key, map) {
+        Ok(index) => Some(&slice[index]),
+        Err(index) => {
+            if index > 0 {
+                Some(&slice[index - 1])
+            } else {
+                None
+            }
+        }
```

Comment:
Just a nit, probably doesn’t matter as the compiler should be smart enough to optimize things away:

slice index syntax can potentially panic. If you use `slice.get(index)` that is both simpler and can potentially generate better code.

```suggestion
        Ok(index) => slice.get(index),
        Err(index) => slice.get(index.checked_sub(1)?),
```